working/2365 - primary constructors/feature-specification.md

```diff
@@ -0,0 +1,510 @@
+   | ...;
+
+<classNamePart> ::= // New rule.
+     'const'? <constructorName> <typeParameters>? <formalParameterList>
+   | <typeWithParameters>;
+
+<classBody> ::= // New rule.
+     '{' (<metadata> <classMemberDeclaration>)* '}'
+   | ';';
+
+<inlineClassDeclaration> ::=
+     'final'? 'inline' 'class' <classNamePart> <interfaces>? <inlineClassBody>;
+
+<inlineClassBody> ::=
+     '{' (<metadata> <inlineMemberDeclaration>)* '}'
+   | ';';
+
+<inlineMemberDeclaration> ::= <classMemberDeclaration>;
+
+<enumType> ::= // Modified rule.
+     'enum' <enumNamePart> <mixins>? <interfaces>? '{'
+        <enumEntry> (',' <enumEntry>)* (',')?
+        (';' (<metadata> <classMemberDeclaration>)*)?
+     '}';
+
+<enumNamePart> ::= // New rule.
+     <constructorName> <typeParameters>? <formalParameterList>
+   | <typeWithParameters>;
+```
+
+The word `inline` is now used in the grammar, but it is not a reserved word
+or a built-in identifier.
+
+### Static processing
+
+A class declaration with a primary constructor *(including `inline`, but
+not `<mixinApplicationClass>`, because the grammar rules do not include
+those)* is desugared to a class declaration without a primary
+constructor. An enum declaration with a primary constructor is desugared
+using the same steps. This determines the static analysis and dynamic
+semantics of a primary constructor.
+
+*In other words, there is no other semantics than the desugaring. When
+desugared, the resulting class declaration is treated exactly the same as
+it would have been if the developer had written the result of the
+desugaring step in the first place.*
+
+A compile-time error occurs if a formal parameter that contains a
+term of the form `this.v`, or `super.v` where `v` is an identifier has the
+modifier `covariant`. *For example, `required covariant int this.v` is an
+error.*
+
+Conversely, it is not an error for the modifier `covariant` to occur on
+other formal parameters of a primary constructor. *This extends the
+existing whitelist of places where `covariant` can occur.*
+
+The desugaring consists of the following steps, where _D_ is the class or
+enum declaration in the program that includes a primary constructor, and
+_D2_ is the result of desugaring. The desugaring step will delete elements
+that amount to the primary constructor, and it will add a new constructor
+_k_.
+
+Where no processing is mentioned below, _D2_ is identical to _D_. Changes
+occur as follows:
+
+The current scope of the formal parameter list of the primary constructor
+in _D_ is the current scope of the class/enum declaration *(in other words,
+the default values cannot see declarations in the class body)*.  Every
+default value in the primary constructor of _D_ is replaced by a fresh
+private name `_n`, and a constant variable named `_n` is added to the
+top-level of the current library. *(This means that we can move the
+parameter declarations including the default value without changing its
+meaning.)*
```

Comment:
This changes the context type of the constant, which will make the following fail:
```dart
class Foo(final double d = 0);
class Bar(final List<int> l = []);
```
The only solution I can see, which doesn't depend on waiting until after type inference, is to copy the type expression of the parameter into the constant variable declaration.

That gives us another problem, when the type depends on type parameters:
```dart
class Foo<T>(final List<T> list = const []);
```
Here we will, eventually, infer a context type of `List<T>` for `const[]`, then infer a type argument of `const <Never>[]`. But that happens so much later that we cannot delay rewriting until we understand what types are. 
(This rewrite probably has to happen before we even validate the type hierarchy, so we don't know what a "type" is yet.)
If we just extract the literal into `const _n = const [];` we get a `List<dynamic>` instead, which is not a valid default value.

We can either do some symbolic processing of the type expression (that's never going to end well), or just admit defeat and say that when we copy the expression, we *include its lexical scope*. Assign a scope to every expression. Or allow some expressions to have lexical scope overrides. (Or whatever we can say which matches what the front-end actually does.)
That assumes the rewriting happens after we have introduced the top-level scope.

But it's not a purely syntactic rewrite. _They never are._

Reply:
Good catch!

I'm now using a rule where the constant variables have a type which is the least closure with respect to such type parameters.